### PR TITLE
New Editor toolbar actions

### LINKED ISSUE #7446: Improve primary editor toolbar actions
### Expected behavior

It should be easier to take primary actions from the top toolbar of the editor. For example, it should be easier to:

* Save as a Draft
* Preview the post
* Toggle HTML mode

### Actual behavior

* To save a draft, I have to go into Post Settings, find Status, and change it to Draft.
* To preview, I have to exit the Editor and tap on Preview in a Posts/Pages list.
* The only way to toggle HTML mode on is in the Formatting Toolbar (not very discoverable)

### Steps to reproduce the behavior

* Open the editor
* See top toolbar :)

### Proposal

editor toolbar updates - android

### Wins

* Keep `Publish/Schedule/Update` button as-is 
* Consolidation puts actions into a single menu, reducing cognitive load
* Brings this inline with what we have on iOS
* Cleans up the UI, leaving more breathing room

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #7446

toolbar_android_final

To test: follow the design as indicated in the issue.
**Notes**: 
1) each of the overflow menu options have separators - couldn't find a way to make the separators appear only after the last of grouped items, even when the items have been actually grouped in the menu xml files as described in the blueprint. I think we can tackle this on a different PR, but let me know otherwise.
2) enabling/disabling overflow menu items such as Undo/Redo doesn't seem to affect the way they look, so we're changing the foreground color programmatically. Enabled buttons are using Android's `Color.BLACK` and disabled buttons are using Android's `Color.GRAY`, but let me know which exact colors should be used for both cases so we can replicate them here.
3) The snackbars are using the default colors for text, except for the `Switched to HTML mode.- UNDO` and `Switched to Visual mode. - UNDO` "UNDO" options for which I'm forcing to use `alert_yellow` (`#f0b849`).
4) the functionality flow is kept as it was, with the only change that `Save as Draft` does actually not only mark the Post as `DRAFT` but also tries to save it to the server as a draft (the half-way implementation I had on Friday only saved the draft locally, which was a difference with how the iOS implementation worked).

cc @iamthomasbishop

### CODE REVIEW

**Comment by mzorz** [comment]
Added the `Publish Now` option replacing the  `Save as Draft` when the Post has already been saved as a Draft.

toolbar_android_final_publish_now_2

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -998,11 +1035,45 @@ public boolean onOptionsItemSelected(final MenuItem item) {
                     mEditPostSettingsFragment.refreshViews();
                 }
                 mViewPager.setCurrentItem(PAGE_SETTINGS);
+            } else if (itemId == R.id.menu_save_as_draft) {
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Nice!  I like the idea, addressed in 0acadb1

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -240,6 +242,8 @@
     // For opening the context menu after permissions have been granted
     private View mMenuView = null;
 
+    private boolean mHtmlModeMenuStateOn = false;
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: addressed in e20e4f7

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 3]
```diff
@@ -690,6 +694,25 @@ private String getSaveButtonText() {
         }
     }
 
+    private String getSaveAsADraftButtonText() {
+        if (!userCanPublishPosts()) {
+            return getString(R.string.submit_for_review);
+        }
+
+        switch (PostStatus.fromPost(mPost)) {
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I meant this idea :) - addressed in 0acadb1

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: ☝️ actually renamed in 6bffe5e

**Comment by mzorz** [comment]
Thanks for the thorough review @malinajirka ! 🙇 

> Ready to publish dialog leaks on rotation

I can’t believe I missed testing orientation changes here! Such a noob thing.

Using the `AppCompatDialogFragment` as per [currently considered best practice](https://developer.android.com/reference/android/app/DialogFragment.html)
I went just a little step ahead and decided to name the FragmentDialog class for this `BaseYesNoFragmentDialog` to make it more general, as seen in 725e4f4 (`added multipurpose BaseYesNoFragmentDialog so we can have orientation-changes-aware dialogs in EditPostActivity`)

In doing this, I found there are at least other 3 occurrences of `AlertDialog.Builder` being used elsewhere in this EditPostActivity that would benefit from using this scheme, so I’ll open a new PR for that after this one gets merged 👍, as it’s a clean and easy change after all.

> Preview post - content disappears on rotation

Let's address this in another issue, I think it's been happening before. Opened https://github.com/wordpress-mobile/WordPress-Android/issues/7483

> We might consider expanding the bottom bar, when the user switches to HTML mode. As the bar is disabled and if the user doesn't realize, it's because of the Html mode, it might a bit confusing. WDYT?

I think it works better now that the status gets propagated between the editor, editor's toolbar, fragment and activity properly (8d6df52) 

> When I enable the HTML mode in the bottom bar and I rotate the device, the icon is not highlighted anymore and the overflow menu displays "Html Mode" instead of "Visual Mode"

> When I enable the HTML mode in the bottom bar, the state is not propagated into the overflow menu (the menu still shows "Html Mode")

Both good catches! 8d6df52 should take care of that

> - Shouldn't the button in the Toolbar be "Publish" when the post is saved as draft (now it's "Update")?

Probably - we decided to not modify anything from the Post statuses cycle except moving the options to the overflow menu, but basically "Update" means that: just update a draft with the most recent content being presented on the screen.

> When the post is published, there is still "Publish Now" option in the overflow menu, which duplicates the "Update" button in the Toolbar -> it might be confusing for the user to have two buttons with the same behaviour but different labels WDYT?

Probably :) - when the Post is published, both "Update" on the action bar and "Publish Now" on the overflow menu update the contents of the Post with the most recent content being displayed on the screen at the moment the option is tapped.

I'll defer to @iamthomasbishop on these last 2 comments, I think we definitely will take these in consideration when we have Aztec/Gutenberg as the only editor and actually get to refactor this piece of code.

I think I covered everything! Thanks a lot really, and ready for another pass! 🙇

**Comment by iamthomasbishop** [comment]
> Shouldn't the button in the Toolbar be "Publish" when the post is saved as draft (now it's "Update")?

Basically what @mzorz said - we are holding off on making structural changes here because it's a complex web we'll need to approach separately :)

> When the post is published, there is still "Publish Now" option in the overflow menu, which duplicates the "Update" button in the Toolbar -> it might be confusing for the user to have two buttons with the same behaviour but different labels WDYT?

I agree - when the post is published, the button in the menu should say `Update`.

---

Long story short, we need to re-think post statuses/actions when working on Gutenberg - it's a messy web. 🕸

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 4]
```diff
@@ -1075,6 +1075,10 @@ public void onClick(View view) {
 
     private void toggleHtmlVisualMode() {
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: 👍 addressed in 92e1b28

**Comment by mzorz** [comment]
> I agree - when the post is published, the button in the menu should say Update

addressed in 9400ca8

**Comment by mzorz** [comment]
> It's ready for merge, when you update the label and fix the CheckStyle issue;).

> Thanks a lot for this PR;)!

Thank you @malinajirka ! :shipit: with your permission  🙇

**Pull request merged**